components/req_cpu/src/linux.rs

```diff
@@ -0,0 +1,690 @@
+use std::marker::PhantomData;
+use std::sync::atomic::Ordering::{AcqRel, Acquire, Relaxed};
+use std::sync::atomic::{AtomicPtr, AtomicU64};
+use std::sync::Arc;
+use std::time::{Duration, Instant};
+
+use collections::{HashMap, HashSet};
+use crossbeam::channel::{unbounded, Receiver, Sender};
+use lazy_static::lazy_static;
+use libc::pid_t;
+use procinfo::pid;
+use procinfo::pid::Stat;
+
+pub fn init_recorder() {
+    lazy_static! {
+        static ref PHANTOM: () = {
+            std::thread::Builder::new()
+                .name("req-cpu-recorder".to_owned())
+                .spawn(move || {
+                    let mut recorder = ReqCpuRecorder::new(RecorderConfig::default());
+
+                    loop {
+                        recorder.handle_collector_registration();
+                        recorder.handle_thread_registration();
+                        recorder.record();
+                        recorder.may_advance_window();
+                        recorder.may_gc();
+
+                        std::thread::sleep(Duration::from_micros(
+                            (recorder.config.record_interval_ms * 1_000.0) as _,
+                        ));
+                    }
+                })
+                .expect("Failed to create recorder thread");
+        };
+    }
+    *PHANTOM
+}
+
+pub fn register_collector(collector: Box<dyn Collector>) -> CollectorHandle {
+    lazy_static! {
+        static ref NEXT_COLLECTOR_ID: Arc<AtomicU64> = Arc::new(AtomicU64::new(1));
+    }
+
+    let id = CollectorId(NEXT_COLLECTOR_ID.fetch_add(1, Relaxed));
+    COLLECTOR_REGISTRATION_CHANNEL
+        .0
+        .send(CollectorRegistrationMsg::Register { collector, id })
+        .ok();
+    CollectorHandle { id }
+}
+
+impl RequestTag {
+    pub fn attach(self: &Arc<Self>) -> Guard {
+        CURRENT_REQ.with(|s| {
+            if s.is_set.get() {
+                panic!("Nested attachment is not allowed.")
+            }
+
+            let prev = s.shared_ptr.swap(self.clone());
+            assert!(prev.is_none());
+            s.is_set.set(true);
+        });
+
+        Guard::default()
+    }
+}
+
+#[derive(Default, Clone)]
+struct SharedReqTagPtr {
+    req_tag: Arc<AtomicPtr<RequestTag>>,
+}
+impl SharedReqTagPtr {
+    fn take(&self) -> Option<Arc<RequestTag>> {
+        let prev_ptr = self.req_tag.swap(std::ptr::null_mut(), Acquire);
+        (!prev_ptr.is_null()).then(|| unsafe { Arc::from_raw(prev_ptr as _) })
+    }
+
+    fn swap(&self, value: Arc<RequestTag>) -> Option<Arc<RequestTag>> {
+        let tag_arc_ptr = Arc::into_raw(value);
+        let prev_ptr = self.req_tag.swap(tag_arc_ptr as _, AcqRel);
+        (!prev_ptr.is_null()).then(|| unsafe { Arc::from_raw(prev_ptr as _) })
+    }
+}
+
+struct LocalReqTag {
+    is_set: Cell<bool>,
+    shared_ptr: SharedReqTagPtr,
+}
+thread_local! {
+    static CURRENT_REQ: LocalReqTag = {
+        let thread_id = unsafe { libc::syscall(libc::SYS_gettid) as libc::pid_t };
```

Comment:
Could you extract real platform-dependent code into a standalone one, and leave the rest shared?